File: output/python/tests/33_doc_history_basic.py
from typing import Any, Optional, List, Dict, Callable

class HistoryBasicFrameEvent:
    def __init__(self, message: str, parameters):
        self._message = message
        self._parameters = parameters


class HistoryBasicFrameContext:
    def __init__(self, event: HistoryBasicFrameEvent, default_return):
        self.event = event
        self._return = default_return
        self._data = {}


class HistoryBasicCompartment:
    def __init__(self, state: str, parent_compartment = None):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment

    def copy(self) -> 'HistoryBasicCompartment':
        c = HistoryBasicCompartment(self.state, self.parent_compartment)
        c.state_args = self.state_args.copy()
        c.state_vars = self.state_vars.copy()
        c.enter_args = self.enter_args.copy()
        c.exit_args = self.exit_args.copy()
        c.forward_event = self.forward_event
        return c


class HistoryBasic:
    def __init__(self):
        self._state_stack = []
        self._context_stack = []
        self.__compartment = HistoryBasicCompartment("A")
        self.__next_compartment = None
        __frame_event = HistoryBasicFrameEvent("$>", None)
        self.__kernel(__frame_event)

    def __kernel(self, __e):
        # Route event to current state
        self.__router(__e)
        # Process any pending transition
        while self.__next_compartment is not None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            # Exit current state
            exit_event = HistoryBasicFrameEvent("<$", self.__compartment.exit_args)
            self.__router(exit_event)
            # Switch to new compartment
            self.__compartment = next_compartment
            # Enter new state (or forward event)
            if next_compartment.forward_event is None:
                enter_event = HistoryBasicFrameEvent("$>", self.__compartment.enter_args)
                self.__router(enter_event)
            else:
                # Forward event to new state
                forward_event = next_compartment.forward_event
                next_compartment.forward_event = None
                if forward_event._message == "$>":
                    # Forwarding enter event - just send it
                    self.__router(forward_event)
                else:
                    # Forwarding other event - send $> first, then forward
                    enter_event = HistoryBasicFrameEvent("$>", self.__compartment.enter_args)
                    self.__router(enter_event)
                    self.__router(forward_event)

    def __router(self, __e):
        state_name = self.__compartment.state
        handler_name = f"_state_{state_name}"
        handler = getattr(self, handler_name, None)
        if handler:
            handler(__e)

    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment

    def gotoC_from_A(self):
        __e = HistoryBasicFrameEvent("gotoC_from_A", None)
        __ctx = HistoryBasicFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        self._context_stack.pop()

    def gotoC_from_B(self):
        __e = HistoryBasicFrameEvent("gotoC_from_B", None)
        __ctx = HistoryBasicFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        self._context_stack.pop()

    def gotoB(self):
        __e = HistoryBasicFrameEvent("gotoB", None)
        __ctx = HistoryBasicFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        self._context_stack.pop()

    def return_back(self):
        __e = HistoryBasicFrameEvent("return_back", None)
        __ctx = HistoryBasicFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        self._context_stack.pop()

    def get_state(self) -> str:
        __e = HistoryBasicFrameEvent("get_state", None)
        __ctx = HistoryBasicFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        return self._context_stack.pop()._return

    def _state_A(self, __e):
        if __e._message == "get_state":
            self._context_stack[-1]._return = "A"
            return
        elif __e._message == "gotoB":
            __compartment = HistoryBasicCompartment("B", parent_compartment=self.__compartment.copy())
            self.__transition(__compartment)
        elif __e._message == "gotoC_from_A":
            self._state_stack.append(self.__compartment.copy())
            __compartment = HistoryBasicCompartment("C", parent_compartment=self.__compartment.copy())
            self.__transition(__compartment)

    def _state_C(self, __e):
        if __e._message == "get_state":
            self._context_stack[-1]._return = "C"
            return
        elif __e._message == "return_back":
            self.__compartment = self._state_stack.pop()
            return

    def _state_B(self, __e):
        if __e._message == "get_state":
            self._context_stack[-1]._return = "B"
            return
        elif __e._message == "gotoC_from_B":
            self._state_stack.append(self.__compartment.copy())
            __compartment = HistoryBasicCompartment("C", parent_compartment=self.__compartment.copy())
            self.__transition(__compartment)


def main():
    print("=== Test 33: Doc History Basic ===")
    h = HistoryBasic()

    # Start in A
    assert h.get_state() == "A", f"Expected 'A', got '{h.get_state()}'"

    # Go to C from A (push A)
    h.gotoC_from_A()
    assert h.get_state() == "C", f"Expected 'C', got '{h.get_state()}'"

    # Return back (pop to A)
    h.return_back()
    assert h.get_state() == "A", f"Expected 'A' after pop, got '{h.get_state()}'"

    # Now go to B
    h.gotoB()
    assert h.get_state() == "B", f"Expected 'B', got '{h.get_state()}'"

    # Go to C from B (push B)
    h.gotoC_from_B()
    assert h.get_state() == "C", f"Expected 'C', got '{h.get_state()}'"

    # Return back (pop to B)
    h.return_back()
    assert h.get_state() == "B", f"Expected 'B' after pop, got '{h.get_state()}'"

    print("PASS: History push/pop works correctly")

if __name__ == '__main__':
    main()

File: output/javascript/tests/33_doc_history_basic.js

// Documentation Example: History with push$/pop$ (History201)

export class HistoryBasicFrameEvent {
    _message;
    _parameters;

    constructor(message, parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}


export class HistoryBasicFrameContext {
    event;
    _return;
    _data;

    constructor(event, default_return) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


export class HistoryBasicCompartment {
    state;
    state_args;
    state_vars;
    enter_args;
    exit_args;
    forward_event;
    parent_compartment;

    constructor(state, parent_compartment = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    copy() {
        const c = new HistoryBasicCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


export class HistoryBasic {
    _state_stack;
    __compartment;
    __next_compartment;
    _context_stack;

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.__compartment = new HistoryBasicCompartment("A");
        this.__next_compartment = null;
        const __frame_event = new HistoryBasicFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    __kernel(__e) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new HistoryBasicFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new HistoryBasicFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new HistoryBasicFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    __router(__e) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = this[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    __transition(next_compartment) {
        this.__next_compartment = next_compartment;
    }

    gotoC_from_A() {
        const __e = new HistoryBasicFrameEvent("gotoC_from_A", null);
        const __ctx = new HistoryBasicFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    gotoC_from_B() {
        const __e = new HistoryBasicFrameEvent("gotoC_from_B", null);
        const __ctx = new HistoryBasicFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    gotoB() {
        const __e = new HistoryBasicFrameEvent("gotoB", null);
        const __ctx = new HistoryBasicFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    return_back() {
        const __e = new HistoryBasicFrameEvent("return_back", null);
        const __ctx = new HistoryBasicFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    get_state() {
        const __e = new HistoryBasicFrameEvent("get_state", null);
        const __ctx = new HistoryBasicFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    _state_A(__e) {
        if (__e._message === "get_state") {
            this._context_stack[this._context_stack.length - 1]._return = "A";
            return;
        } else if (__e._message === "gotoB") {
            const __compartment = new HistoryBasicCompartment("B", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        } else if (__e._message === "gotoC_from_A") {
            this._state_stack.push(this.__compartment.copy());
            const __compartment = new HistoryBasicCompartment("C", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        }
    }

    _state_B(__e) {
        if (__e._message === "get_state") {
            this._context_stack[this._context_stack.length - 1]._return = "B";
            return;
        } else if (__e._message === "gotoC_from_B") {
            this._state_stack.push(this.__compartment.copy());
            const __compartment = new HistoryBasicCompartment("C", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        }
    }

    _state_C(__e) {
        if (__e._message === "get_state") {
            this._context_stack[this._context_stack.length - 1]._return = "C";
            return;
        } else if (__e._message === "return_back") {
            this.__transition(this._state_stack.pop());
            return;
        }
    }
}

function main() {
    console.log("=== Test 33: Doc History Basic ===");
    const h = new HistoryBasic();

    // Start in A
    if (h.get_state() !== "A") {
        throw new Error(`Expected 'A', got '${h.get_state()}'`);
    }

    // Go to C from A (push A)
    h.gotoC_from_A();
    if (h.get_state() !== "C") {
        throw new Error(`Expected 'C', got '${h.get_state()}'`);
    }

    // Return back (pop to A)
    h.return_back();
    if (h.get_state() !== "A") {
        throw new Error(`Expected 'A' after pop, got '${h.get_state()}'`);
    }

    // Now go to B
    h.gotoB();
    if (h.get_state() !== "B") {
        throw new Error(`Expected 'B', got '${h.get_state()}'`);
    }

    // Go to C from B (push B)
    h.gotoC_from_B();
    if (h.get_state() !== "C") {
        throw new Error(`Expected 'C', got '${h.get_state()}'`);
    }

    // Return back (pop to B)
    h.return_back();
    if (h.get_state() !== "B") {
        throw new Error(`Expected 'B' after pop, got '${h.get_state()}'`);
    }

    console.log("PASS: History push/pop works correctly");
}

main();
